samcli/lib/build/build_strategy.py

```diff
@@ -0,0 +1,240 @@
+        """
+        Build the unique definition and then copy the artifact to the corresponding layer folder
+        """
+        layer = layer_definition.layer
+        LOG.info("Building layer '%s'", layer.name)
+        if layer.build_method is None:
+            raise MissingBuildMethodException(
+                f"Layer {layer.name} cannot be build without BuildMethod. Please provide BuildMethod in Metadata.")
+        return {layer.name: self._build_layer(layer.name,
+                                              layer.codeuri,
+                                              layer.build_method,
+                                              layer.compatible_runtimes)}
+
+
+class CachedBuildStrategy(BuildStrategy):
+    """
+    Cached implementation of Build Strategy
+    For each function and layer, it first checks if there is a valid cache, and if there is, it copies from previous
+    build. If caching is invalid, it builds function or layer from scratch and updates cache folder and md5 of the
+    function or layer.
+    For actual building, it uses delegate implementation
+    """
+
+    def __init__(self, build_graph, delegate_build_strategy, base_dir, build_dir, cache_dir,
+                 is_building_specific_resource):
+        super().__init__(build_graph)
+        self._delegate_build_strategy = delegate_build_strategy
+        self._base_dir = base_dir
+        self._build_dir = build_dir
+        self._cache_dir = cache_dir
+        self._is_building_specific_resource = is_building_specific_resource
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._clean_redundant_cached()
+
+    def build(self):
+        result = {}
+        with self, self._delegate_build_strategy:
+            result.update(super().build())
+        return result
+
+    def build_single_function_definition(self, build_definition):
+        """
+        Builds single function definition with caching
+        """
+        code_dir = str(pathlib.Path(self._base_dir, build_definition.codeuri).resolve())
+        source_md5 = dir_checksum(code_dir)
+        cache_function_dir = pathlib.Path(self._cache_dir, build_definition.get_uuid())
+        function_build_results = {}
+
+        if not cache_function_dir.exists() or build_definition.get_source_md5() != source_md5:
+            LOG.info("Cache is invalid, running build and copying resources to function build definition of %s",
+                     build_definition.get_uuid())
+            build_result = self._delegate_build_strategy.build_single_function_definition(build_definition)
+            function_build_results.update(build_result)
+
+            if cache_function_dir.exists():
+                shutil.rmtree(str(cache_function_dir))
+
+            build_definition.set_source_md5(source_md5)
+            for _, value in build_result.items():
+                osutils.copytree(value, cache_function_dir)
+                break
```

Comment:
Why do we need the `for` if we're `break`ing on the first result?

Reply:
The build_result is a dictionary with items like this {'func1': 'func1Dir', 'func2': 'func2Dir', ...}. All the content in the func1Dir, func2Dir, ... are all the same since they have the same build content. Here I just want to copy content for any one of them to the cache_function_dir. Other option will be change this for-loop to:
```python
build_function_dir = list(build_result.values())[0]
osutils.copytree(build_function_dir, cache_function_dir)
```

Reply:
I see! In that case either is fine (maybe I'd even prefer the `for` as it avoids `IndexError`s). A one-line comment would definitely be helpful, as it's not immediately obvious _why_ we're using such a pattern.